src/control-plane/backend/lambda/listen-stack-status/index.ts

```
@@ -0,0 +1,222 @@
+import { logger } from '../../../../common/powertools';
+import { aws_sdk_client_common_config, marshallOptions, unmarshallOptions } from '../../../../common/sdk-client-config';
+import { WorkflowParallelBranch, WorkflowState, WorkflowStateType } from '../api/common/types';
+
+const ddbClient = new DynamoDBClient({
+  ...aws_sdk_client_common_config,
+  region: process.env.AWS_REGION,
+});
+
+const docClient = DynamoDBDocumentClient.from(ddbClient, {
+  marshallOptions: { ...marshallOptions },
+  unmarshallOptions: { ...unmarshallOptions },
+});
+
+const clickStreamTableName = process.env.CLICKSTREAM_TABLE_NAME ?? '';
+const prefixTimeGSIName = process.env.PREFIX_TIME_GSI_NAME ?? '';
+
+export interface CloudFormationStackStatusChangeNotificationEventDetail {
+  'stack-id': string;
+  'status-details': {
+    status: StackStatus;
+    'status-reason': string;
+  };
+}
+
+export const handler = async (event: SQSEvent): Promise<void> => {
+  const eventBody = event.Records[0].body;
+  const eventBridgeEvent = JSON.parse(eventBody) as EventBridgeEvent<'CloudFormation Stack Status Change', CloudFormationStackStatusChangeNotificationEventDetail>;
+  const eventDetail = eventBridgeEvent.detail;
+  const region = eventBridgeEvent.region;
+  const stackId = eventDetail['stack-id'];
+  const stackName = stackId.split('/')[1];
+
+  if (!stackName.startsWith('Clickstream')) {
+    return;
+  }
+  const stackDetail = await describeStack(stackId, region);
+  if (!stackDetail) {
+    logger.error('Failed to describe stack: ', { stackId, region });
+    return;
+  }
+  if (stackDetail.ParentId) {
+    return;
+  }
+  logger.info('Detail: ', { stackName: stackName, status: stackDetail.StackStatus });
+
+  const pipelineId = getPipelineIdFromStackName(stackName);
+  const pipeline = await getPipeline(pipelineId);
+  if (!pipeline) {
+    logger.error('Failed to get pipeline by pipelineId: ', { pipelineId });
+    return;
```

This swallows this event. I prefer to throw the exception to retry it later.